### PR TITLE
feat: add parseAssistantMessageV3 with robust XML parsing

### PR DESCRIPTION
**Pull request by samhvw8**
### Related GitHub Issue

Closes: #1317  #3039 #4426 #3404 

### Description

- Implement parseAssistantMessageV3 using partial-xml-stream-parser library
- Handle CDATA sections properly for safe parsing of XML-like content
- Support mixed HTML and tool call parsing without corruption
- Add stopNodes configuration for efficient parsing of large content blocks
- Improve partial message detection for streaming responses

The new parser resolves issues with:
- XML-like content in write_to_file breaking the parser
- Mixed HTML tags being incorrectly parsed as tool calls
- Better handling of incomplete/streaming messages
- More reliable extraction of tool parameters from complex structures

Tests added for edge cases including nested XML, CDATA sections, and mixed HTML/tool content.

### Type of Change

- [ ] 🐛 **Bug Fix**: Non-breaking change that fixes an issue.
- [ ] ✨ **New Feature**: Non-breaking change that adds functionality.
- [ ] 💥 **Breaking Change**: Fix or feature that would cause existing functionality to not work as expected.
- [ ] ♻️ **Refactor**: Code change that neither fixes a bug nor adds a feature.
- [ ] 💅 **Style**: Changes that do not affect the meaning of the code (white-space, formatting, etc.).
- [ ] 📚 **Documentation**: Updates to documentation files.
- [ ] ⚙️ **Build/CI**: Changes to the build process or CI configuration.
- [ ] 🧹 **Chore**: Other changes that don't modify `src` or test files.

### Pre-Submission Checklist

- [ ] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [ ] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [ ] **Self-Review**: I have performed a thorough self-review of my code.
- [ ] **Code Quality**:
    - [ ] My code adheres to the project's style guidelines.
    - [ ] There are no new linting errors or warnings (`npm run lint`).
    - [ ] All debug code (e.g., `console.log`) has been removed.
- [ ] **Testing**:
    - [ ] New and/or updated tests have been added to cover my changes.
    - [ ] All tests pass locally (`npm test`).
    - [ ] The application builds successfully with my changes.
- [ ] **Branch Hygiene**: My branch is up-to-date (rebased) with the `main` branch.
- [ ] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [ ] **Changeset**: A changeset has been created using `npm run changeset` if this PR includes user-facing changes or dependency updates.
- [ ] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### CODE REVIEW

**Comment by samhvw8** [comment]
benchmark for this 

| Test Case                                         | V1 Time (ms) | V2 Time (ms) | V3 Time (ms) | V1/V2 Ratio | V1/V3 Ratio | V2/V3 Ratio | V1 Heap (bytes) | V2 Heap (bytes) | V3 Heap (bytes) |
| ------------------------------------------------- | ------------ | ------------ | ------------ | ----------- | ----------- | ----------- | ---------------- | ---------------- | ---------------- |
| Simple text message                               |       0.0361 |       0.0191 |       0.0048 |        1.90 |        7.53 |        3.97 |              274 |           28,603 |            7,578 |
| Message with a simple tool use                    |       0.0453 |       0.0268 |       0.0093 |        1.69 |        4.85 |        2.87 |           17,051 |            5,990 |           13,358 |
| Message with a complex tool use (write_to_file)   |       0.0931 |       0.0460 |       0.0121 |        2.02 |        7.72 |        3.81 |           69,888 |           33,721 |           13,514 |
| Message with multiple tool uses                   |       0.1454 |       0.0473 |       0.0148 |        3.08 |        9.80 |        3.18 |           22,503 |           54,283 |           19,022 |
| Large message with repeated tool uses             |       5.8215 |       1.3892 |       0.2035 |        4.19 |       28.61 |        6.83 |          270,203 |           33,269 |           38,020 |

File: src/core/assistant-message/parseAssistantMessageV3.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+			"browser_action.text",
+		],
+		allowedRootNodes: ["thinking", ...toolNames],
+	})
+
+	const contentBlocks: AssistantMessageContent[] = []
+	let isPartial = false
+
+	// Process the message and capture intermediate results
+	const parseResult = parser.parseStream(assistantMessage)
```
[the remark this answers is not in the document]
Comment (samhvw8) [reply src/core/assistant-message/parseAssistantMessageV3.ts:78]: Already handle in parser 💪🏻

**Comment by samhvw8** [comment]
@daniel-lxs this pr need an follow up pr to clean all parse step for nested xml in tool, this new parser will handle all nest xml in parse assistant message function

**Comment by AugLuk** [comment]
Nice work!
[Issue 3039](https://github.com/RooCodeInc/Roo-Code/issues/3039) also talks about rogue + symbols in apply_diff. Can I fix that part, or are you already working on it?

**Comment by samhvw8** [comment]
> Nice work! [Issue 3039](https://github.com/RooCodeInc/Roo-Code/issues/3039) also talks about rogue + symbols in apply_diff. Can I fix that part, or are you already working on it?

I put this pr in draft why i testing it and fixing it, i believe that will be fix on my xml apply diff pr, but we need test it again after that merged

**Comment by samhvw8** [comment]
@daniel-lxs  fixed, now it work correctly :D

**Comment by d-oit** [comment]
@samhvw8 
https://github.com/RooCodeInc/Roo-Code/issues/3404
[roo_task_jun-7-2025_7-59-37-pm.md](https://github.com/user-attachments/files/20640041/roo_task_jun-7-2025_7-59-37-pm.md)

image

**Comment by samhvw8** [comment]
> @samhvw8 #3404 [roo_task_jun-7-2025_7-59-37-pm.md](https://github.com/user-attachments/files/20640041/roo_task_jun-7-2025_7-59-37-pm.md)
> 
> image

hmm let me check it again

**Comment by samhvw8** [comment]
@d-oit  @daniel-lxs you can test it again <3, thanks :D

**Comment by samhvw8** [comment]
performance after fix issues :D 

| Test Case                                         | V1 Time (ms) | V2 Time (ms) | V3 Time (ms) | V1/V2 Ratio | V1/V3 Ratio | V2/V3 Ratio | V1 Heap (bytes) | V2 Heap (bytes) | V3 Heap (bytes) |
| ------------------------------------------------- | ------------ | ------------ | ------------ | ----------- | ----------- | ----------- | ---------------- | ---------------- | ---------------- |
| Simple text message                               |       0.0364 |       0.0179 |       0.0062 |        2.03 |        5.88 |        2.90 |           40,271 |           28,525 |            9,634 |
| Message with a simple tool use                    |       0.0459 |       0.0266 |       0.0131 |        1.73 |        3.51 |        2.04 |           19,375 |            5,982 |           16,006 |
| Message with a complex tool use (write_to_file)   |       0.0925 |       0.0420 |       0.0153 |        2.20 |        6.06 |        2.75 |           70,352 |           33,666 |           18,329 |
| Message with multiple tool uses                   |       0.1472 |       0.0488 |       0.0225 |        3.01 |        6.54 |        2.17 |            7,326 |           30,189 |           31,568 |
| Large message with repeated tool uses             |       5.5114 |       1.3154 |       0.3341 |        4.19 |       16.50 |        3.94 |          276,965 |           30,642 |           92,694 |

**Comment by samhvw8** [comment]
@KJ7LNW @mrubens  this pr introduce streaming parsing xml (parse incomplete xml) this can improve our UX (stream data when it incomplete -- partial mode) (this work even on nested xml like insert_content and read_file, ... )

**Comment by d-oit** [comment]
> @d-oit @daniel-lxs you can test it again <3, thanks :D

- Test with your latest commit. 
- works now in ask mode directly :thumbsup: 
but not from another mode (orchestrator) as starting task point. 
markdown task:
[roo_task_jun-9-2025_10-09-47-am_orchestrator.md](https://github.com/user-attachments/files/20651440/roo_task_jun-9-2025_10-09-47-am_orchestrator.md)

image
image


- a few other test cases: 
[parseAssistantMessageV3.test.txt](https://github.com/user-attachments/files/20651493/parseAssistantMessageV3.test.txt)

Unfortunately, I don't have time to test any further. I'll stick with my solution for now – if desired, my issue can be closed with this as the solution:
Modify the system prompt to instruct the AI to escape special characters:

When using ask_followup_question, ensure that suggestion content escapes XML special characters:
- Replace < with &lt;
- Replace > with &gt;  
- Replace & with &amp;

**Comment by samhvw8** [comment]
@d-oit just test again using sonnet 3.5, in orchestrator mode it work perfectly, i suspect that it is gemini issues
image

image

image

File: src/core/assistant-message/parseAssistantMessageV3.ts [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+	const textNodeNameKey = "#text"
+	const attributeNamePrefixKey = "@"
+	const parser = new PartialXMLStreamParser({
+		textNodeName: textNodeNameKey,
+		attributeNamePrefix: attributeNamePrefixKey,
+		alwaysCreateTextNode: true,
+		parsePrimitives: false,
+		// Removing maxDepth to allow proper parsing of nested structures
+		// maxDepth: 1, // this for backward compatible with old version (old version only support 2 levels (1 level depth))
+		stopNodes: [
```
Comment (KJ7LNW) [inline src/core/assistant-message/parseAssistantMessageV3.ts:16]: I am not sure if it should be part of this pull request or part of a tool class refactor, but having tool-specific language in a generic parser implementation somehow seems inappropriate.

It would be nice if each tool can register what it expects to be a stopNode

File: src/core/parser/constants.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+export const STATIC_OPENING_TAG_REGEX = /<\s*([\w:-]+)([^>]*?)(\/?)\>/
+export const STATIC_CLOSING_TAG_REGEX = /<\/\s*([\w:-]+)\s*>/
+
+export const COMMON_ENTITIES: Record<string, string> = {
+	lt: "<",
+	gt: ">",
+	amp: "&",
+	quot: '"',
+	apos: "'",
+}
```
Comment (KJ7LNW) [inline src/core/parser/constants.ts:10]: I do not think they should be hard coded, it should probably use a proper existing library for handling html entities in general.

**Review by KJ7LNW — COMMENTED** [review]
does this support attributes like `<tag attr="value">` ?

**Comment by samhvw8** [comment]
> does this support attributes like `<tag attr="value">` ?

Current version -> Yes, but i think i will remove it for now :(

File: src/core/assistant-message/parseAssistantMessageV3.ts (hunk 2 above)
Comment (KJ7LNW) [reply src/core/assistant-message/parseAssistantMessageV3.ts:16]: also if you need to iterate `stopNodes` regularly, then it should probably be a set or a dictionary so that you do not have to do a linear scan. 

for example, if you have to check `stopNodes` for every new token response from the model while it streams, then a linear scan (`O(n)`) will be slower compared to a dictionary or set lookup ( `O(log(n))` ).

**Comment by hannesrudolph** [comment]
Closed in favour of #4866
